Make pin_memory/is_pinned use BackendSelect

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60547 Make pin_memory/is_pinned use BackendSelect**

These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

ezyang committed:
Make pin_memory/is_pinned use BackendSelect

These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang commented:
cc @sujoysaraswati

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3229,9 +3229,14 @@
 
 - func: is_pinned(Tensor self, Device? device=None) -> bool
   variants: method
+  dispatch:
```
Does everything just work when we change an op from being math-composite (`CompositeImplicitAutograd`) to being backend-specific?

My first thought was that we'd need to add an entry to derivatives.yaml, but I see some other ops that are backend-specific but do not have a derivatives.yaml entry (e.g. `aten::is_set_to`), and it looks like the autograd codegen still creates dummy autograd kernels for them.

ezyang replied:
I checked the generated code for is_pinned and it looks fine:

```
bool is_pinned(c10::DispatchKeySet ks, const at::Tensor & self, c10::optional<at::Device> device) {
  auto& self_ = unpack(self, "self", 0);
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::is_pinned(ks & c10::after_autograd_keyset, self_, device);
  })();
  auto result = _tmp;
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with is_pinned that does not support it.");
  return result;
} 
```

These functions are weird because they don't have any meaningful autograd content (as they don't return a Tensor). Effectively, all it does exclude autograd key and then redispatch. I think this is just the AD codegen being smart enough to not error if it sees the function doesn't return a Tensor. cc @albanD for any further thoughts.

albanD replied:
This is exactly what happens: we only generate actual code for ops that have either one differentiable input or output: https://github.com/pytorch/pytorch/blob/b7298f499d6a88bf2da92d4f07c6f4a52e73adf4/tools/autograd/gen_variable_type.py#L461

Note that if you have differentiable input/output but you don't specify anything in derivatives.yaml, it is equivalent to doing:
```
- name: my_fn(Tensor foo, Tensor bar, int baz) -> Tensor
  foo: not_implemented("my_fn")
  bar: not_implemented("my_fn")
```

SujoysGithub replied:
I have a question on how this would work for a non CUDA backend. Since the empty_cpu() uses a pinned memory allocator for CUDA alone, are the other backends explicitly required to call a tensor.pin_memory() afterwards?

ezyang replied:
Yes. And then we can fix this if we change `pin_memory` argument on empty to take a DEVICE rather than a bool, but that is a very complicated change and very BC breaking without very careful handling so you shouldn't count on it happening in the near future.

bdhirsh commented on aten/src/ATen/templates/RegisterBackendSelect.cpp:
```diff
@@ -15,8 +15,32 @@ namespace {
+
 TORCH_LIBRARY_IMPL(aten, BackendSelect, m) {
   ${backend_select_function_registrations};
+  m.impl("is_pinned", is_pinned);
```
It looks like the reason you hardcoded these is that the codegen only generates backend select kernels for ops with a `TensorOptions` arg. Is your thinking that manually writing the code here is worth the save on confusion from having people pass in a `TensorOptions` arg that ignores everything but the device?

I could see that, but probably worth an explicit comment at least

ezyang replied:
Yeah, I much prefer only a device argument, all the other stuff is completely unnecessary. It honestly didn't occur to me that anyone would possibly want to do this, which is why there's no comment.

bdhirsh replied:
ok fair 😛

bdhirsh commented on aten/src/ATen/templates/RegisterBackendSelect.cpp:
```diff
@@ -15,8 +15,32 @@ namespace {
+bool is_pinned(const Tensor& self, c10::optional<at::Device> device) {
+  // Only CPU tensors can be pinned
+  if (!self.is_cpu()) {
+    return false;
```
Kinda confused by this. Won't this mean that if you call `is_pinned(my_tensor, kCuda)`, we'll just short-circuit here and return false instead of dispatching into `is_pinned_cuda`?

Also, if you want any short-circuiting here for not-allowed devices, isn't that redundant to the `is_pinned_default` kernel?

bdhirsh replied:
oh, do we need this because of static dispatch? We want `is_pinned(_, kCPU)` to return false, not throw an error. Although another option would be to register `is_pinned_default` to both CPU and CompositeExplicitAutograd - not sure if that's better though.

ezyang replied:
Look carefully: the test is on `self.is_cpu()`, not `device`. If you pass a CUDA tensor to is pinned, we always return false, because the pinning concept is taking CPU memory and putting it in a place where the CUDA driver can read it out directly. So in your example, so long as `my_tensor` is CPU, we will not short circuit here.

bdhirsh replied:
bleh yeah, that makes sense. thanks!

ezyang commented:
cc @bhosmer

bdhirsh reviewed (approved):
thanks for the clarifications!

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2208,3 +2208,6 @@
 - name: segment_reduce(Tensor data, str reduce, *, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False, Scalar? initial=None) -> Tensor
   data: _segment_reduce_backward(grad, result, data, reduce, lengths)
+
+- name: pin_memory(Tensor(a) self, Device? device=None) -> Tensor(a)
```
pin_memory is a view op?
If it is a no-op in some cases (returning self as is), this is not ok for autograd. You will have to keep this one as composite and use another function to do the pinning that always return different memory.

ezyang replied:
whoops. OK yes will need to fix this. I wonder if any tests gonna catch this...

albanD replied:
https://github.com/pytorch/pytorch/pull/60286 should once it lands haha
If you want to test this function only, you can give an input that requires grad and is already pinned, that should fail as well.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM!

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang commented:
cc @albanD I redid this PR to exactly preserve old autograd semantics

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).

ezyang committed:
Update on "Make pin_memory/is_pinned use BackendSelect"


These now dispatch on the optional Device argument, which specifies
what device you want to pin for.  We now directly register pinned
memory implementations for CUDA specifically, eliminating the need
for extra virtual methods.

This makes it possible for other backends to override the behavior
of pinned memory, c.f. https://github.com/pytorch/pytorch/pull/59291

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331881](https://our.internmc.facebook.com/intern/diff/D29331881)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331881).
